File: parse_questions.py
import re
import json

with open("extracted_text_2.txt", "r") as f:
    text = f.read()

questions = []

idx = text.rfind("Correct Answers")
questions_text = text[:idx]
answers_text = text[idx:]

# parse answers
answer_map = {}
answer_matches = re.findall(r'(\d+)\.\s*([A-E])', answers_text.replace('\n', ' '))
for match in answer_matches:
    answer_map[int(match[0])] = match[1]

# parse questions
question_blocks = re.split(r'Question \d+', questions_text)[1:]

for i, block in enumerate(question_blocks):
    q_num = i + 1
    
    positions = {}
    for opt in ['A.', 'B.', 'C.', 'D.', 'E.']:
        pos = block.find(opt)
        if pos != -1:
            positions[opt] = pos
            
    if not positions:
        continue
        
    sorted_opts = sorted(positions.items(), key=lambda x: x[1])
    
    question_text = block[:sorted_opts[0][1]].strip()
    question_text = question_text.replace('\ufb01', 'fi').replace('\ufb02', 'fl').replace('\x0c', '')
    
    options = {}
    for j in range(len(sorted_opts)):
        opt_letter = sorted_opts[j][0][0]
        start = sorted_opts[j][1] + 2 # skip "A."
        if j < len(sorted_opts) - 1:
            end = sorted_opts[j+1][1]
        else:
            end = len(block)
        
        opt_text = block[start:end].strip()
        opt_text = opt_text.replace('\ufb01', 'fi').replace('\ufb02', 'fl').replace('\x0c', '')
        options[opt_letter] = opt_text
        
    questions.append({
        "id": q_num,
        "question": question_text,
        "options": options,
        "answer": answer_map.get(q_num, "")
    })

with open("questions.js", "w") as f:
    f.write("const examData = " + json.dumps(questions, indent=2) + ";\n")

print(f"Parsed {len(questions)} questions")

File: scripts/parse-questions.ts
/**
 * Local helper to turn an extracted-text exam dump into the JSON array the app
 * consumes. Port of the old parse_questions.py.
 *
 * Usage:
 *   npm run parse -- <input.txt> [output.json]
 *   (defaults: extracted_text_2.txt -> questions.json)
 */
import { readFileSync, writeFileSync } from 'node:fs';

interface Question {
  id: number;
  question: string;
  options: Record<string, string>;
  answer: string;
}

const inputPath = process.argv[2] ?? 'extracted_text_2.txt';
const outputPath = process.argv[3] ?? 'questions.json';

const clean = (s: string): string =>
  s.replace(/ﬁ/g, 'fi').replace(/ﬂ/g, 'fl').replace(/\x0c/g, '');

const text = readFileSync(inputPath, 'utf-8');

const idx = text.lastIndexOf('Correct Answers');
const questionsText = idx === -1 ? text : text.slice(0, idx);
const answersText = idx === -1 ? '' : text.slice(idx);

// Parse the answer key: "12. B" style entries.
const answerMap = new Map<number, string>();
const answerRe = /(\d+)\.\s*([A-E])/g;
const flatAnswers = answersText.replace(/\n/g, ' ');
for (const m of flatAnswers.matchAll(answerRe)) {
  answerMap.set(Number(m[1]), m[2]);
}

// Split into question blocks on "Question N".
const blocks = questionsText.split(/Question \d+/).slice(1);
const optionLabels = ['A.', 'B.', 'C.', 'D.', 'E.'];

const questions: Question[] = [];

blocks.forEach((block, i) => {
  const qNum = i + 1;

  const positions: Array<[string, number]> = [];
  for (const opt of optionLabels) {
    const pos = block.indexOf(opt);
    if (pos !== -1) positions.push([opt, pos]);
  }
  if (positions.length === 0) return;

  positions.sort((a, b) => a[1] - b[1]);

  const questionText = clean(block.slice(0, positions[0][1]).trim());

  const options: Record<string, string> = {};
  for (let j = 0; j < positions.length; j++) {
    const letter = positions[j][0][0];
    const start = positions[j][1] + 2; // skip "A."
    const end = j < positions.length - 1 ? positions[j + 1][1] : block.length;
    options[letter] = clean(block.slice(start, end).trim());
  }

  questions.push({
    id: qNum,
    question: questionText,
    options,
    answer: answerMap.get(qNum) ?? '',
  });
});

writeFileSync(outputPath, JSON.stringify(questions, null, 2) + '\n');
console.log(`Parsed ${questions.length} questions -> ${outputPath}`);
